File: php/protein-translation/ProteinTranslation.php
<?php

declare(strict_types=1);

class ProteinTranslation
{
    private const CODON_TO_PROTEIN = [
        "AUG" => "Methionine",
        "UUU" => "Phenylalanine",
        "UUC" => "Phenylalanine",
        "UUA" => "Leucine",
        "UUG" => "Leucine",
        "UCU" => "Serine",
        "UCC" => "Serine",
        "UCA" => "Serine",
        "UCG" => "Serine",
        "UAU" => "Tyrosine",
        "UAC" => "Tyrosine",
        "UGU" => "Cysteine",
        "UGC" => "Cysteine",
        "UGG" => "Tryptophan",
        "UAA" => self::STOP,
        "UAG" => self::STOP,
        "UGA" => self::STOP,
    ];
    private const INVALID_CODON = "not really a codon :D";
    private const STOP = "STOP";

    public function getProteins(string $codons): array
    {
        $proteins = [];

        foreach(str_split($codons, 3) as $codon) {
            $protein = static::CODON_TO_PROTEIN[$codon] ?? static::INVALID_CODON;
            if ($protein === static::STOP) {
                break;
            } else if ($protein === static::INVALID_CODON) {
                throw new InvalidArgumentException("Invalid codon");
            } else {
                $proteins[] = $protein;
            }
        }

        return $proteins;
    }
}

File: go/protein-translation/protein_translation.go
package proteintranslation

import "errors"

var ErrStop = errors.New("ErrStop")
var ErrInvalidBase = errors.New("ErrInvalidBase")

func FromRNA(rna string) ([]string, error) {

	var protein []string

	for {
		if len(rna) == 0 {
			return protein, nil
		}

		if len(rna) < 3 {
			return nil, ErrInvalidBase
		}

		codon := rna[:3]
		aminoAcid, err := FromCodon(codon)

		if err == ErrInvalidBase {
			return nil, err
		}
		if err == ErrStop {
			return protein, nil
		}

		protein = append(protein, aminoAcid)
		rna = rna[3:]

	}
}

func FromCodon(codon string) (string, error) {
	switch codon {
	case "AUG":
		return "Methionine", nil
	case "UUU", "UUC":
		return "Phenylalanine", nil
	case "UUA", "UUG":
		return "Leucine", nil
	case "UCU", "UCC", "UCA", "UCG":
		return "Serine", nil
	case "UAU", "UAC":
		return "Tyrosine", nil
	case "UGU", "UGC":
		return "Cysteine", nil
	case "UGG":
		return "Tryptophan", nil
	case "UAA", "UAG", "UGA":
		return "", ErrStop
	default:
		return "", ErrInvalidBase
	}
}
